src/ripple/basics/base_uint.h

```diff
@@ -363,99 +373,62 @@ class base_uint
     }
 
     /** Parse a hex string into a base_uint
-        The string must contain exactly bytes * 2 hex characters and must not
-        have any leading or trailing whitespace.
-    */
-    bool
-    SetHexExact(const char* psz)
-    {
-        unsigned char* pOut = begin();
 
-        for (int i = 0; i < sizeof(data_); ++i)
-        {
-            auto hi = charUnHex(*psz++);
-            if (hi == -1)
-                return false;
-
-            auto lo = charUnHex(*psz++);
-            if (lo == -1)
-                return false;
-
-            *pOut++ = (hi << 4) | lo;
-        }
+        The input must be precisely `2 * bytes` hexadecimal characters
+        long, with one exception: the value '0'.
 
-        // We've consumed exactly as many bytes as we needed at this point
-        // so we should be at the end of the string.
-        return (*psz == 0);
-    }
-
-    /** Parse a hex string into a base_uint
-        The input can be:
-            - shorter than the full hex representation by not including leading
-              zeroes.
-            - longer than the full hex representation in which case leading
-              bytes are discarded.
-
-        When finished parsing, the string must be fully consumed with only a
-        null terminator remaining.
-
-        When bStrict is false, the parsing is done in non-strict mode, and, if
-        present, leading whitespace and the 0x prefix will be skipped.
-    */
-    bool
-    SetHex(const char* psz, bool bStrict = false)
+        @param sv A null-terminated string of hexadecimal characters
+        @return true if the input was parsed properly; false otherwise.
+     */
+    [[nodiscard]] bool
+    parseHex(std::string_view sv)
     {
-        // Find beginning.
-        auto pBegin = reinterpret_cast<const unsigned char*>(psz);
-        // skip leading spaces
-        if (!bStrict)
-            while (isspace(*pBegin))
-                pBegin++;
+        zero();
 
-        // skip 0x
-        if (!bStrict && pBegin[0] == '0' && tolower(pBegin[1]) == 'x')
-            pBegin += 2;
+        if (sv == "0")
+            return true;
 
-        // Find end.
-        auto pEnd = pBegin;
-        while (charUnHex(*pEnd) != -1)
-            pEnd++;
+        if (sv.size() != bytes * 2)
+            return false;
 
-        // Take only last digits of over long string.
-        if ((unsigned int)(pEnd - pBegin) > 2 * size())
-            pBegin = pEnd - 2 * size();
+        auto out = data();
 
-        unsigned char* pOut = end() - ((pEnd - pBegin + 1) / 2);
+        auto parseHexChar = [](char c) {
+            if (auto x = charUnHex(c); x != -1)
+                return static_cast<unsigned char>(x);
 
-        *this = beast::zero;
-
-        if ((pEnd - pBegin) & 1)
-            *pOut++ = charUnHex(*pBegin++);
+            throw c;
```

Comment:
I like this improved API for `base_uint`, but I don't recommend using exceptions for flow control.  It is expensive both in terms of performance and code size.  Furthermore the `static_cast` is in the wrong place to be effective since the expression `(hi << 4) + lo` promotes everything back to `int`.

Here is what I recommend:

```c++
    [[nodiscard]] bool
    parseHex(std::string_view sv)
    {
        zero();

        if (sv == "0")
            return true;

        if (sv.size() != bytes * 2)
            return false;

        auto out = data();

        auto in = sv.begin();

        while (in != sv.end())
        {
            auto const hi = charUnHex(*in++);
            auto const lo = charUnHex(*in++);
            if (hi == -1 || lo == -1)
                return false;
            *out++ = static_cast<value_type>((hi << 4) + lo);
        }

        return true;
    }
```